axlearn/common/optimizers.py

```diff
@@ -1993,3 +2015,93 @@ def _update2(u: Tensor, param: OptParam):
+            context.
+
+    This function returns a new `PartitionedGradientTransformation` that
+    1. Puts matched states of the wrapped optimizer on `offload_dst` through the partition function
+       during state initialization in the trainer.
+    2. Copies the matched states to `offload_src` before `optimizer.update` is called.
+    3. Copies the matched updated states to `offload_dst` after `optimizer.update` is called.
+
+    The regex pattern is matched against the full path of each optimizer state. An example full
+    path is optimizer/1/0/mu/decoder/transformer/repeat/layer/feed_forward/linear1_0. If the
+    pattern should not depend on model structure, you can use ".*mu.*" to offload all `mu`.
+
+    The .update function of the returned `PartitionedGradientTransformation` must be called within
+    a jit function.
+
+    Example usage:
+    ```python
+    your_opt = adamw_optimizer(...)
+    offloaded_opt = offload_optimizer(your_opt)
+    ```
+
+    When using `skip_and_clip_by_global_norm` with this offload optimizer, you must wrap the entire
+    `skip_and_clip_by_global_norm` inside. Do not wrap the inner of `skip_and_clip_by_global_norm`
+    or you will get errors. Correct example:
+    ```
+    offloaded_opt = offload_optimizer(skip_and_clip_by_global_norm(inner=adamw_optimizer(...)))
+    ```
+    The reason is that `skip_and_clip_by_global_norm` conditionally chooses the previous optimizer
+    state and the updated new optimizer state using `jnp.where`, which doesn't support tensors on
+    `pinned_host` memory space.
+    """
+    optimizer = maybe_instantiate(optimizer)
+    if offload_src is None or offload_dst is None:
+        raise ValueError(
+            "offload_src and offload_dst cannot be None when using optimizer offloading."
+        )
+
+    logging.info("Optimizer offloading from %s to %s enabled.", offload_src, offload_dst)
+
+    def init_fn(params: NestedOptParam):
+        return optimizer.init(params)
+
+    def _move_fn(state: optax.OptState, dst: MemoryKind) -> optax.OptState:
+        # TransferToMemoryKind let us change the memory kind of tensors without specifying the full
+        # sharding (i.e. jax.sharding.NamedSharding). Although there's no documentation about it,
+        # it's specified in the API signature. Reference:
+        # https://github.com/jax-ml/jax/blob/21f8885a9e104b8828c9a8b721eed0c68b622691/jax/_src/api.py#L2220
+        return jax.tree.map(
+            lambda path, tensor: jax.device_put(tensor, TransferToMemoryKind(dst))
+            if re.fullmatch(pattern, path)
+            else tensor,
+            tree_paths(state),
+            state,
+        )
+
+    def update_fn(updates: optax.Updates, state: optax.OptState, params: NestedOptParam):
+        state = _move_fn(state, offload_src)
+        updates, state = optimizer.update(updates, state, params)
+        state = _move_fn(state, offload_dst)
```

Comment:
I wonder whether the explicit `device_put` calls mean that we need to move all optimizer states into device before optimizer computation, leading to high device memory usage.

In theory, optimizer computation can be streaming---the optimizer states of variables can be updated separately so we can stream them into and out of device memory as needed. How do we achieve this?

An alternative is to use `with_sharding_constraint` calls instead of `device_put` calls, where the partition specs given to the `with_sharding_constraint` calls will put the arrays in the device memory. WDYT?

Reply:
I've verified in profile before that `device_put` won't move everything at once. When we pass a pytree of arrays to `device_put`, each array in the pytree is moved independent of one another. The exact order is decided by the latency hiding scheduler. The scheduler will try to overlap the transfers of each state with the state update on TPU whenever possible. There is some memory spike due the the temporary state in HBM, but the spike is much less than the full memory usage of all states. Moreover, when the optimizer is run, all activations are released, so we have less memory pressure at that point in time.

If I remove clip by global norm and use `StackedTransformerLayer`, the optimizer transfers can be overlapped with the backward pass, leading to faster step time and less memory spike. The reason is that clip by global norm requires all gradients to be computed before optimizer can be executed, and `RepeatedTransformerLayer` stack weights of all layers into a single buffer which does not allow the layer-by-layer optimizer execution.

Reply:
Sure, will add a comment. Regarding using `with_sharding_constraint`, I would prefer not to use that as I don't want to populate the full sharding spec when alternative methods (i.e. `device_put`) that can modify the memory kind directly exist.

Reply:
Done.